### DISCUSSION TITLE
Is the Airtable library compatible with edge-runtime, and thus Cloudflare Pages?

### DISCUSSION THREAD

**Discussion by komali2** [vercel/next.js#44843] [Help] [answered]
We’re trying to get a next.js on Cloudflare pages, and most things are happy, except for trying to invoke Airtable.js library functions: https://github.com/airtable/airtable.js/ from within next.js page files.

Our concern is that the airtable package isn’t in whatever “subset” of NPM packages are available when using nextjs’ “edge-runtime,” a requirement per Cloudflare of non-statically rendered next.js sites: https://nextjs.org/docs/api-reference/edge-runtime

However we can’t find a definitive list of what packages are compatible. We do notice that unless we enable the edge runtime per page, it 404s when we deploy to Cloudflare pages. Each page that can’t be statically rendered requires:

```
export const config = {
   runtime: 'experimental-edge',
 };
```

However, if such a page imports Airtable, then, that page crashes next.js’ server with errors about malformed relative URLs. Very weird errors but we think maybe blocks of incompatible with edge-runtime code are being dropped from the airtable library.

For now we’re poking through the library line by line to see if we can find anything that the edge runtime documentation (above) explicitly calls out as unimplemented, but it’d be nice if anyone here could expand and let us know if we’re going down the wrong path or perhaps just tell us if this is impossible / would require forking the airtable library and using implemented APIs.

**Comment by icyJoseph** [comment, marked as answer]
I can start by saying that they do `typeof window !== 'undefined'` checks to either use Web APIs or Node.js API - which is a problem because AFAIK, edge runtimes, do not have a window, but are not quite a Node.js environment either.

I am not saying that's the root of your issue, but it certainly means that wherever those checks are made, your edge runtime functions are going the Node.js paths, and if the API's they expect are not there, well, all hell breaks loose.

**Comment by komali2** [comment]
See also: https://community.cloudflare.com/t/is-the-airtable-js-library-compatible-with-cloudflare-pages/452308

In short, probably not, because of as mentioned by icyjoseph incompatible APIs used, but also, the fact that environment variables aren't really available "server-side". 

See also: https://github.com/cloudflare/next-on-pages/issues/1
